hazelcast-jet-core/src/main/java/com/hazelcast/jet/windowing/SlidingWindowP.java

```diff
@@ -100,34 +102,38 @@ protected boolean tryProcessPunc0(@Nonnull Punctuation punc) {
     }
 
     private Traverser<Object> slidingWindowTraverser(Punctuation punc) {
-        Traverser<Object> traverser = Traversers.traverseStream(
-                range(nextFrameSeqToEmit, nextFrameSeqToEmit = wDef.higherFrameSeq(punc.seq()), wDef.frameLength())
-                        .mapToObj(frameSeq ->
-                                computeWindow(frameSeq, seqToKeyToFrame.remove(frameSeq - wDef.windowLength()))
-                                        .entrySet().stream()
-                                        .map(e -> new Frame<>(frameSeq, e.getKey(), finishF.apply(e.getValue()))))
-                        .flatMap(identity()));
-
-        traverser = traverser.append(punc);
-        return traverser;
+        long rangeStart = nextFrameSeqToEmit;
+        nextFrameSeqToEmit = wDef.higherFrameSeq(punc.seq());
+        return Traversers.traverseStream(range(rangeStart, nextFrameSeqToEmit, wDef.frameLength()).boxed())
+                .flatMap(frameSeq -> Traversers.traverseIterable(computeWindow(frameSeq).entrySet())
+                        .map(e -> (Object) new Frame<>(frameSeq, e.getKey(), finishF.apply(e.getValue())))
+                        .onFirstNull(() -> completeWindow(frameSeq)))
+                .append(punc);
     }
 
-    private Map<K, F> computeWindow(long frameSeq, Map<K, F> evictedFrame) {
+    private void completeWindow(long frameSeq) {
+        Map<K, F> evictedFrame = seqToKeyToFrame.remove(frameSeq - wDef.windowLength() + wDef.frameLength());
         if (deductF != null) {
-            // add leading-edge frame
```

Comment:
While trying to validate your claims on memory savings, i ran this code:
```java
public static void main(String[] args) throws Exception {
    SlidingWindowPTest t = new SlidingWindowPTest();
    t.hasDeduct = true;
    t.isMutableFrame = true;
    t.before();
    SlidingWindowP<Object, ?, Long> proc = t.processor;
    for (long frameSeq = 0; frameSeq < 40; frameSeq++) {
        Punctuation punc = new Punctuation(frameSeq);
        while (!proc.tryProcessPunc0(punc)) {
        }
        for (long key = 0; key < 200_000; key++) {
            Frame<String, MutableLong> frame = new Frame<>(frameSeq, "" + key, MutableLong.valueOf(1));
            while (!proc.tryProcess0(frame)) {
            }
        }
    }
    System.out.println("seqToKeyToFrame " + proc.seqToKeyToFrame.size());
}

private static class MockOutbox implements Outbox {

    @Override
    public int bucketCount() {
        return 1;
    }

    @Override
    public boolean offer(int ordinal, @Nonnull Object item) {
        System.out.println("offer " + ordinal + ' ' + item);
        return true;
    }

    @Override
    public boolean offer(int[] ordinals, @Nonnull Object item) {
        return true;
    }
}

@Before
public void before() {
    WindowDefinition windowDef = new WindowDefinition(1, 0, 1);
    WindowOperation<Object, ?, Long> operation;

    if (isMutableFrame) {
        operation = WindowOperation.<Object, MutableLong, Long>of(
                MutableLong::new,
                (acc, val) -> {
                    throw new UnsupportedOperationException();
                },
                (acc1, acc2) -> {
                    acc1.value += acc2.value;
                    return acc1;
                },
                hasDeduct ? (acc1, acc2) -> {
                    acc1.value -= acc2.value;
                    return acc1;
                } : null,
                acc -> acc.value);
    } else {
        operation = WindowOperation.of(
                () -> 0L,
                (acc, val) -> {
                    throw new UnsupportedOperationException();
                },
                (acc1, acc2) -> acc1 + acc2,
                hasDeduct ? (acc1, acc2) -> (Long) acc1 - acc2 : null,
                identity());
    }

    processor = slidingWindow(windowDef, operation).get();
    processor.init(new MockOutbox(), mock(Context.class));
}
```

It turns out that there's a memory leak and no frames are evicted. I ran the same code on master and frames were being evicted.